guide/src/python_from_rust.md

```diff
@@ -230,3 +230,30 @@ def leaky_relu(x, slope=0.01):
 
 [`Python::run`]: https://docs.rs/pyo3/latest/pyo3/struct.Python.html#method.run
 [`py_run!`]: https://docs.rs/pyo3/latest/pyo3/macro.py_run.html
+
+## Need to use a context manager from Rust?
+
+Use context managers by directly invoking `__enter__` and `__exit__`.
+
+```rust
+use pyo3::prelude::*;
+use pyo3::types::PyModule;
+
+fn main() {   
+    Python::with_gil(|py| {
+        let custom_manager = PyModule::from_code(py, r#"
+class House(object):
+    def __init__(self, address):
+        self.address = address
+    def __enter__(self):
+        print(f"Welcome to {self.address}!")
+    def __exit__(self, type, value, traceback):
+        print(f"Thank you for visiting {self.address}, come again soon!")
+        "#, "objects.py", "objects").unwrap();
+        
+        let house = custom_manager.call1("House", ("123 Main Street",)).unwrap();
+        house.call_method0("__enter__").unwrap();
+        house.call_method1("__exit__", ("", "", "")).unwrap();
```

Comment:
So in Python the arguments to `__exit__` are the exception information in the case of an error occurring inside the `with` block.

When working from Rust we'll need to set this ourselves - we should either use `py.None()` if no error occurred, or the exception tuple in the case where there was an error.

... we in theory could get even fancier than this - according to the [docs for `object.__exit__()`](https://docs.python.org/3/reference/datamodel.html#object.__exit__) if `__exit__` returns a truthy value then we should suppress the original exception, otherwise we should raise it.

E.g.

```suggestion
        let result = do_some_python_operation_that_might_fail();
        match result {
            Ok(_) => {
                let none = py.None();
                house.call_method1("__exit__", (none, none, none)).unwrap();
            },
            Err(e) => house.call_method1(
                "__exit__",
                (e.ptype(py), e.pvalue(py), e.ptraceback(py))
            ).unwrap();
        }
```

the "re-raising" in the case of a falsy value would be better demonstrated in a `#[pyfunction]` where the `Err` can be returned back to Python.